### PR TITLE
Feat: auto update ODS data files

### PR DESCRIPTION
**Pull request by cindyli**
This pull request checks if a new data file of COVID-19 assessment centre locations is published at the [Ontario Data Service
website](https://data.ontario.ca/dataset/covid-19-assessment-centre-locations). If there is one, this script downloads
the file, updates the corresponding latest.json and issues a pull request against [the COVID data repository](https://github.com/inclusive-design/covid-assessment-centres/).

### CODE REVIEW

File: scripts/fetchODCDataFilesUtils.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		try {
+			let res = await axios.get(dataSourceURL);
+			let dom = new JSDOM(res.data);
+
+			// find the CSV download link
+			let downloadLink, publishedDate;
+			let as = dom.window.document.querySelectorAll("a.dataset-download-link");
+			for (let a of as) {
+				const link = a.getAttribute("href");
+				if (link.slice(-4) === ".csv") {
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:37]: link.endsWith() is a bit more idiomatic here

File: tests/fetchODCDataFilesUtilsTests.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			input: undefined,
+			expected: "NaN-NaN-NaN"
+		},
+		special2: {
+			message: "The input is empty",
+			input: "",
+			expected: "NaN-NaN-NaN"
+		},
+		special3: {
+			message: "The input format is invalid",
```
Comment (amb26) [inline tests/fetchODCDataFilesUtilsTests.js:106]: Should add some more tests here to check marginal cases such as partially specified or otherwise out of range dates such as "20 November", "53 November", "November 2020" etc.

File: scripts/fetchODCDataFilesUtils.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+					downloadLink = link;
+					var dateElmContent = a.closest(".resource-item").querySelector(".description.details").innerHTML;
+					const matchStartStr = "Last Updated: ";
+					const matchEndStr = " |";
+					publishedDate = dateElmContent.substring(dateElmContent.lastIndexOf(matchStartStr) + matchStartStr.length, dateElmContent.lastIndexOf(matchEndStr));
+					publishedDate = module.exports.formatDate(publishedDate);
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:43]: We should probably signal an error at this point if the date is invalid - and also ensure that there really is a single uniquely determinable invalid date output from this function

File: scripts/fetchODCDataFilesUtils.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+			let downloadLink, publishedDate;
+			let as = dom.window.document.querySelectorAll("a.dataset-download-link");
+			for (let a of as) {
+				const link = a.getAttribute("href");
+				if (link.slice(-4) === ".csv") {
+					downloadLink = link;
+					var dateElmContent = a.closest(".resource-item").querySelector(".description.details").innerHTML;
+					const matchStartStr = "Last Updated: ";
+					const matchEndStr = " |";
+					publishedDate = dateElmContent.substring(dateElmContent.lastIndexOf(matchStartStr) + matchStartStr.length, dateElmContent.lastIndexOf(matchEndStr));
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:42]: I'm not terribly fond of them myself, but I expect a regex such as "Last Updated: (.*) |" might be more intentionally clear to readers, and perhaps easier to externalise as an option too

File: scripts/fetchODCDataFilesUtils.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+					publishedDate = module.exports.formatDate(publishedDate);
+					break;
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:44]: Rather than bailing out, I suggest this should be an as.filter() returning all matching links, and then a warning issued if we end up finding more than one - or is the idea that we expect to use only the first of any matching set?

File: scripts/fetchODCDataFilesUtils.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	 * Download a file from the given download URL and write into a target local file.
+	 * @param {String} downloadURL The download URL
+	 * @param {String} targetFileLocation The target file location including the path and file name
+	 */
+	downloadDataFile: async (downloadURL, targetFileLocation) => {
+		try {
+			let res = await axios.get(downloadURL);
+			fs.writeFileSync(targetFileLocation, res.data, "utf8");
+		} catch (error) {
+			// do nothing
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:99]: Why swallow error rather than propagating it?

File: scripts/fetchODCDataFilesUtils.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+					break;
+				}
+			}
+
+			return {
+				downloadUrl: downloadLink,
+				publishedDate: publishedDate
+			};
+		} catch (error) {
+			console.log("Error in getDataSource(): " + error);
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:53]: Rather than have this side-effect, better to just propagate the error (either as exception or rejection) and let the driver determine how to log it

File: scripts/fetchODCDataFilesUtils.js [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+				}
+			  }
+			}`
+		}, {
+			headers:headers
+		});
+
+		if (createPRResponse.data.errors) {
+			console.log("Error at creating the new pull request: ", createPRResponse.data.errors);
+			process.exit(1);
```
Comment (amb26) [inline scripts/fetchODCDataFilesUtils.js:167]: Again, rather than side-effect, better to raise an error payload and let the driver determine what to do with it

File: tests/fetchODCDataFilesUtilsTests.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		},
+		expected: {
+			downloadUrl: undefined,
+			publishedDate: undefined
+		}
+	}
+};
+
+fluid.each(getDataSourceTestCases, function (oneCase, key) {
+	jqUnit.asyncTest("Test getDataSource() - " + key, function () {
```
Comment (amb26) [inline tests/fetchODCDataFilesUtilsTests.js:64]: If you update to 3.0.0-dev.20201113T133151Z.fdbc62672.FLUID-6145 you should be able to use the new idiomatic support for promise-based tests - e.g.

    jqUnit.test("Test getDataSource() - " + key, function () {
        return utils.getDataSource(fullSitePath).then(function (result) {
			jqUnit.assertDeepEq(oneCase.message, oneCase.expected, result);
    };
)

File: scripts/fetchODCDataFiles.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+if (process.argv.length < 3) {
+	console.log("Usage: node scripts/fetchODCDataFiles.js dataSourceURL dataDirInRepo\n");
+	console.log("dataSourceURL: The URL to webpage where the data file for COVID-19 assessment centre locations is published. An example: https://data.ontario.ca/dataset/covid-19-assessment-centre-locations");
+	console.log("dataDirInRepo: The directory where the downloaded data file should be placed in the COVID-19 data repository(https://github.com/inclusive-design/covid-assessment-centres/). This directory is relative to the root directory of the repository ");
+	process.exit(1);
+}
+
+// The central COVID-19 data repository on Github
+const covidDataRepoUrl = "https://github.com/cindyli/covid-assessment-centres.git";
```
Comment (ptcher) [inline scripts/fetchODCDataFiles.js:47]: Should this URL be hardcoded? I remember when we first spoke about this feature we agreed to endpoints like this should be lifted into a JSON configuration file. Same might be appropriate for the dataSourceURL and dataDirInRepo above.

**Review by ptcher — COMMENTED** [review]
This looks great. The top-level script particularly seems very sensibly structured to me.

File: scripts/fetchODCDataFilesUtils.js (hunk 5 above)
Comment (cindyli) [reply scripts/fetchODCDataFilesUtils.js:44]: I will improve the code to report an error when more than one csv download links are found or no csv link is found.

File: tests/fetchODCDataFilesUtilsTests.js (hunk 9 above)
Comment (cindyli) [reply tests/fetchODCDataFilesUtilsTests.js:64]: @amb26, the suggested change doesn't work.

After some digging, it seems 3.0.0-dev.20201113T133151Z.fdbc62672.FLUID-6145 release doesn't include your latest jqUnit support for promise based tests.

If you search "FLUID-6577" in [the jqUnit.js at the commit hash fdbc62672](https://github.com/amb26/infusion/blob/fdbc62672406586c2399d213a2e4e1909334d7bc/tests/test-core/jqUnit/js/jqUnit.js), nothing will be found.

I was expecting jqUnit.js changes in this merge: https://github.com/fluid-project/infusion/commit/1619f14b960ccf2727db0db45570811186a26aeb

Let me know if I missed anything. Thanks.

Comment (amb26) [reply tests/fetchODCDataFilesUtilsTests.js:64]: Very sorry, the release seemed to be from the right date since I couldn't think of another reason for a release around that time. I've republished 3.0.0-dev.20201113T153152Z.32176dcbe.FLUID-6145 from a commit 2 hours later which does include the fix!

Comment (cindyli) [reply tests/fetchODCDataFilesUtilsTests.js]: Thanks for pointing out, @amb26. This release works perfectly. I should have looked more carefully thru the infusion release list. ;)

**Comment by cindyli** [comment]
@amb26, @ptcher, thanks for reviews. All comments have been addressed and ready for another review.

File: tests/fetchODCDataFilesUtilsTests.js (hunk 9 above)
Comment (amb26) [reply tests/fetchODCDataFilesUtilsTests.js]: No, it was entirely my fault - fluid-publish dates the release from the commit rather from the current time and the release didn't exist at the time you looked for it. Somehow at the time I had managed to make the release from the previous commit.

**Comment by amb26** [comment]
Thanks, I checked this out and it works great - oddly when running the script I get the following warning: 
````
(node:38984) DeprecationWarning: Instance method `debug.destroy()` is deprecated and no longer does anything. It will be removed in the next major version of `d
ebug`.
````
I assume this is some node method which is being called by one of our dependencies but I was unable to find any hits anywhere for any part of this message.

**Comment by cindyli** [comment]
> Thanks, I checked this out and it works great - oddly when running the script I get the following warning:
> 
> ```
> (node:38984) DeprecationWarning: Instance method `debug.destroy()` is deprecated and no longer does anything. It will be removed in the next major version of `d
> ebug`.
> ```
> 
> I assume this is some node method which is being called by one of our dependencies but I was unable to find any hits anywhere for any part of this message.

My trace yesterday did pinpoint to the dependency of `simple-git` as the source that throws this warning. Unfortunately I didn't find a way to get rid of it.

Another thing that puzzles me is, `simple-git` has been used and executed from a couple of weeks ago, but this warning started to show up only from yesterday. I didn't figure out what was changed and triggered this report.

The good thing is, it's a bit annoying but doesn't affect the script result.

Thanks for the merge, @amb26.
